utils/xpdfopen/pdfopen.1
.\" This manpage is licensed under the GNU Public License
.TH PDFOPEN 1 2014-05-16  "pdfopen 0.84" ""
.
.SH NAME
pdfopen, pdfclose \- open or close a PDF file viewer
.
.SH SYNOPSIS
.B pdfopen 
[\fIOPTION\fP] 
\fIFILE.PDF\fP
.br
.B pdfclose
\fIFILE.PDF\fP
.
.SH OPTIONS
.TP 8
.B -h, --help
output help and exit
.TP
.B -v, --version
output the version number and exit
.TP
.B -p <n>
when starting the viewer, display page <n> of the document (notes: not
all viewers support this feature, and even in those cases this option
only has effect when the document is initially loaded)
.TP
.B -r, --reset_focus
after sending commands to the PDF viewer, attempt to reset the input focus to
the window which had focus before the commands were sent
.TP
.B -viewer
<\fIar9\fP|\fIar9-tab\fP|\fIar8\fP|\fIar7\fP|\fIar5\fP|\fIxpdf\fP|\fIevince\fP>
use (respectively) Adobe Reader 9 (in a new window), Adobe Reader 9
(in a new tab of a running AR9, if any),
Adobe Reader 8, Adobe Reader 7, Adobe Reader 5, xpdf or evince as the
PDF viewer program.
Adobe Reader 9 (in a new window) is the default.
.
.SH RATIONALE
At certain points of TeX document preparation, many people repeat a
"edit-compile-view" cycle.  Since PDF viewers such as Adobe's Acrobat
Reader ("\fBacroread\fP") do not automatically refresh the display when the
PDF file changes, this cycle can be more cumbersome than desired.
The
.B pdfopen
program provides the ability to automate the reloading of the PDF
document when it is changed.

Note: there seems to be little need for
.B pdfclose
under GNU/Linux, since (unlike the situation for MS windows)
.B acroread
does not lock the PDF file, which would prevent
.B pdftex
(or a DVI to PDF converter) from creating a new version of the PDF
output file.  However,
.B pdfclose
is provided in case someone finds it useful.
.
.SH DESCRIPTION
.B pdfopen
searches for an instance of the specified (or default) PDF viewer
displaying the specified PDF file.
If there is already an instance of the given viewer displaying the
given file, the viewer is instructed to reload the file.  If no such
instance is found, \fBpdfopen\fP attempts to run the specified viewer
on the specified document.

The default viewer is "\fBacroread\fP", which could start any one
of a number of versions of Acrobat Reader, depending on what is
installed on your system.  However, the commands to reload the current
document vary from one version of \fBacroread\fP to another;
consequently, if you are using a version of \fBacroread\fP other than
AR9, you should explicitly specify the viewer program.

Version 0.84 of \fBpdfopen\fP accepts the following viewer options:
.br
\fIar9\fP, \fIar9-tab\fP, \fIar8\fP, \fIar7\fP, \fIar5\fP, \fIxpdf\fP,
and \fIevince\fP.
.br
The difference between \fIar9\fP and \fIar9-tab\fP is significant
when there is no instance of AR9 already displaying the requested
document.  In this situation, while \fIar9\fP will
request \fBacroread\fP to create a new instance of \fBacroread\fP
(and thus open a new window) by using the \fB-openInNewInstance\fP
argument, \fIar9-tab\fP starts \fBacroread\fP without this argument;
if there is already an instance of \fBacroread\fP running, a new
tab will be opened in an existing window.

.B pdfclose
searches for one of the above PDF viewers displaying the given file
and instructs the viewer to "close" the window.  In most cases, the
PDF viewer continues to run, possibly now displaying just a blank
window.  (This behaviour varies somewhat from one PDF viewer to another.)
.
.SH PORTABILITY AND AVAILABILITY
These programs have been tested on Slackware64 Version 14.1 and a few
other versions / distributions of GNU/Linux.  The code is reasonably
generic and should work out of the box using most recent X11
implementations.  (Reports to the contrary are welcome, particularly
if they come with robust fixes.)

These programs are designed for X11-based systems.  If you somehow
find compiled versions of these programs on a system using another
window system, they are very unlikely to be of any use to you.

Source and binaries of the programs can be downloaded from 
CTAN://support/xpdfopen/
(e.g., http://mirror.ctan.org/support/xpdfopen).
.
.SH CAVEATS
If you use
.I ar9-tab
to reload the PDF document and the instance
of
.B acroread
with the given document is currently displaying some other document,
the command causes your document to be displayed, but not reloaded.

.B pdfopen
works by looking for a window with a name (window title) matching that
expected for the given viewer and document.  If for some reason your
viewer's window name is not as expected,
.B pdfopen
may not work for you.

With at least AR9 and some window managers, using
.B pdfopen
to reload the document gives focus to the
.B acroread
window, even though the mouse cursor is not necessarily in that
window.  This can be annoying.  The \fB-reset_focus\fP option can be
used to deal with this problem.
.
.SH AUTHOR 
This manual page was written by Jim Diamond <Jim.Diamond@acadiau.ca>.
I am the current maintainer of the X11 versions of
.B pdfopen
and
\fBpdfclose\fP.
Report any bugs you find to me.  Feature enhancement requests are
welcome, coded enhancements even more so.

Past authors:
Fabrice Popineau wrote the MS-windows versions of
.B pdfopen
and
.B pdfclose
upon which these programs were originally based.
Taco Hoekwater created the GNU/Linux versions, up to Version 0.61
(including some documentation which inspired parts of this man page).
Peter Breitenlohner has contributed both code and suggestions to
versions later than 0.61.


ChangeLog
2014-04-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Makefile.am: New names for README.* files.

2014-04-17  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2014-04-04  Karl Berry  <karl@tug.org>

	* Build (TL_MAKE,MAKE): set both in all cases.
	(buildenv.log): include pwd, and use .log instead of .txt.

2014-04-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-04-01.

2014-01-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-01-01.

2013-12-26  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (ACLOCAL): Ensure aclocal.m4 is up to date.

2013-10-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* doc: New directory for texlive.info & Co.
	* Makefile.am, configure.ac: Adapted.

2013-10-05  Karl Berry  <karl@tug.org>

	* Build: reorder variables and settings for clarity.
	(TL_CONF_BANNER): restore variable, default is empty string.
	(TL_POSTCONFIGURE): new variable, default true.

2013-09-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Drop ACLOCAL_AMFLAGS.
	* configure.ac: Use AC_CONFIG_MACRO_DIRS.
	Use KPSE_RECURSE_LIBS in m4/kpse-setup.m4 to recurse into all
	TeX specific libraries.

2013-09-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-09-01.

	* Makefile.am: Moved definition of cf_silent* from here ...
	* am/recurse.am: ... to here 

2013-08-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Add the make target 'texlinks'.

2013-07-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/rebuild.am: Avoid still more 'parallel make' problems.

2013-07-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/rebuild.am: Avoid more 'parallel make' problems.

2013-07-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/*.am: Moved Makefile fragments from libs/am/ and tekk/am/.
	* am/rebuild.am: New Makefile fragment to rebuild a library.
	* am/reconfig.am: New Makefile fragment to reconfigure.
	* am/recurse.am: New Makefile fragment to recurse into subdirs.

	* Makefile.am, configure.ac: Use am/recurse.am.

2013-06-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Added Automake options 'tar-pax' for paths
	longer than 100 chars and 'no-dist-gzip'.
	* version.ac: Switch 2013 => 2014/dev'
	* tardate.ac: Switch to 2013-06-24.

2013-03-26  Karl Berry  <karl@tug.org>

	* Build: use $MAKE rather than $TL_MAKE for the make clean
	in the unlikely event of a local Makefile.

2013-03-21  Mojca Miklavec  <mojca.miklavec.lists@gmail.com>

	* Build: Remove GNUMAKE variable (after freetype2 as the only library
	which needed it has been fixed).

2013-03-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-03-01.

2013-02-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Building for Windows does not require 'ln -s'.
	* README.config: Document '--without-ln-s'.

2013-02-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (mandir, infodir): Change texmf => texmf-dist.

2013-01-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-01-24.
	* Makefile.am: Running texlinks in texk/texlive/tl_scripts.

2013-01-18  Karl Berry  <karl@tug.org>

	* Build: remove use of time.

2012-12-17  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Enable silent rules.

2012-11-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-11-16.

2012-10-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-10-24.

2012-08-19  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-08-19.

2012-08-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Running texlinks in texk/texlive/linked_scripts.

2012-07-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Build: Add TL_COMPILER_GFLAGS also to OBJCXXFLAGS.

2012-07-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2012 => 2013/dev.

2012-06-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-06-15.

2012-05-21  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Test for GNU make before creating Makefile,
	required for FreeBSD make and perhaps others.

2012-05-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README: Update the list of prerequisites.
	* README.config: Mention configure options for biber.

2012-04-12  Karl Berry  <karl@tug.org>

	* Build (TL_COMPILER_GFLAGS): new envvar, used to override -g
	for all three compilers, with Build -g.

2012-04-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-04-03.

2012-03-31  Karl Berry  <karl@tug.org>

	* Build (TL_CONF_LARGEFILE): switch default to --enable-largefile.
	More info in README.config.

2012-03-12  Karl Berry  <karl@tug.org>

	* README.config: 32-bit rather than 32Bit

2012-03-11  Karl Berry  <karl@tug.org>

	* README.config: text tweaks for poppler/libxpdf/luatex/etc.

2012-01-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Report `$enable_largefile'.

2012-01-13  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-01-13.

2011-12-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Replace DISTCHECK_CONFIGURE_FLAGS by
	AM_DISTCHECK_CONFIGURE_FLAGS (for automake-1.11.2).

2011-12-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Use TL specific installation directories only
	for $enable_native_texlive_build=yes (also datarootdir).
	Display the installation directories.

	* README.config, README.distro: Adapted.

2011-10-05  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-10-05.

2011-08-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2011-08-10.

2011-07-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am [EXTRA_DIST]: Add m4/ChangeLog.

2011-07-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2011 => 2012/dev.

2011-06-21  Karl Berry  <karl@tug.org>

	* Build: report final exit status.

2011-04-15  Peter Breitenlohner  <peb@mppmu.mpg.de>


utils/xpdfopen/pdfopen.c
/*
 * pdfopen.c.
 *
 * Heavily modified by Jim Diamond (jim.diamond@acadiau.ca) 2010/04/11
 * to work with AR 9, xpdf and evince.
 * (Also some mods by Peter Breitenlohner <tex-live@tug.org> from 2009
 * and suggestions from 2010, and some suggestions from Karl Berry
 * (2011).)
 *
 * Note 1: the calling sequence was changed considerably to match what
 * the program actually seemed to do as well as what made sense to me.
 *
 * Note 2: if there are two instances of acroread displaying that file,
 * only one is reloaded.
 *
 * Further modified 2010/05/03 to work with ar5, and to include some code
 * that does *not* work with okular (see note below).
 *
 * Modified 2014/05/16 (V 0.84) to add capability to open a pdf file
 * at a specified page (currently working with AR9, xpdf and evince).
 * 
 * Note 3: Unlike xpdf and AR[5,7,8,9], okular will not accept X events
 * from this program unless the okular window has focus.  On my system
 * (Slackware64 13.0, using fvwm2), the functions I added to sendx.c to try
 * to change the focus to the okular window do not work.  (See also the
 * comments in sendx.c.)  If anyone cares to contribute working code,
 * please send it to me for inclusion.
 */

#include    <stdio.h>
#include    <stdlib.h>
#include    <string.h>
#include    <unistd.h>
#include    <sys/types.h>
#include    <sys/wait.h>

#include    "sendx.h"
#include    "xpdfopen.h"
#include    "utils.h"

#define	    DEFINE_EXTERNS
#include    "externs.h"

#define     MAX_RELOAD_CMDS		10
#define     MAX_PRE_PAGE_OPTIONS	 2	/* See struct defn below */
#define     MAX_POST_PAGE_OPTIONS	 1	/* See struct defn below */
	       					/* Overkill ... */
#define     MAX_OPTIONS			(10 + MAX_PRE_PAGE_OPTIONS \
                                         + MAX_POST_PAGE_OPTIONS)

/* A string unlikely to be used as a real command by a PDF viewer */
#define     FORK_EXEC_CMD	"*FORK EXEC*"


/* Used in the pdf_viewer_t type... see comment below */
typedef enum
{
    USE_SENDX, USE_FORK_EXEC
} reload_method_t;


/*
 * This struct contains the information needed for both
 * (a) initially starting up a PDF viewer, and
 * (b) telling the viewer to reload the document.
 * Of course, the viewer has to support that functionality.
 * Two possible ways are supported.
 * (a) Send the viewer's window (a) simulated keypress(ss)
 * (b) fork()/exec() a second instance of the program which somehow signals
 *     the running instance.  Note that this is problematical if the
 *     arguments for the "reload" case are different than the options
 *     for the "start up the viewer" case, because the command-line
 *     (and environment variable) syntax have to be extended.
 *     Arguably, if both the start-up and reload can be done with shell
 *     commands, the user should just write a shell script and be done with
 *     it.
 * These could be extended with this as well:
 * (c) send a Unix signal to the currently running viewer.
 *     This would require augmenting reload_method_t and pdf_viewer_t.
 */

typedef struct
{
    const char * short_name;			/* cmd line option name */
    const char * program_name;			/* program to exec() */
    reload_method_t method;			/* method to induce reload */
    const char * window_name;			/* when no file is viewed */
    const char * window_name_fmt;		/* fmt string for file viewed */
    const char * exec_pre_options[MAX_OPTIONS];	/* before filename in cmd */
    const char * exec_post_options[MAX_OPTIONS];/* after filename in cmd */
    const char * reload_cmds[MAX_RELOAD_CMDS];	/* cmds to reload a file */
    /*
     * If the page number goes before the filename, use these options.
     * The last one must be a printf format string with a %d and
     * no other conversions specs.
     * End list with a NULL.
     */
    const char * initial_page_pre_options[MAX_PRE_PAGE_OPTIONS + 1]; /* */
    /*
     * If the page number goes after the filename, use these options.
     * The last one must be a printf format string with a %d and
     * no other conversions specs.
     * End list with a NULL.
     */
    const char * initial_page_post_options[MAX_POST_PAGE_OPTIONS + 1]; /* */
} pdf_viewer_t;


/* Reload commands (or prefixes thereof) */
#define     CTRL	"Ctrl-"
#define     ALT		"Alt-"
#define     CTRL_ALT	"Ctrl-Alt-"
#define     FN		"Fn-"
#define     FOCUS_IN	"FOCUS-IN"
#define     FOCUS_OUT	"FOCUS-OUT"



/*
 * The first struct represents the default viewer with the default options.
 */
static
pdf_viewer_t pdf_viewers[] =
{
    {"ar9", "acroread", USE_SENDX,
     AR9_WIN_NAME, AR9_WIN_NAME_FMT,
     {"-openInNewInstance", NULL},
     {NULL},
     {"Ctrl-R", NULL},
     {"/a", "page=%d", NULL},
     {NULL}
    },
    {"ar9-tab", "acroread", USE_SENDX, 
     AR9_WIN_NAME, AR9_WIN_NAME_FMT,
     {NULL},
     {NULL},
     {"Ctrl-R", NULL},
     {NULL},
     {NULL}
    },
    {"ar8", "acroread", USE_SENDX,
     AR8_WIN_NAME, AR8_WIN_NAME_FMT,
     {NULL},
     {NULL},
     {"Ctrl-W", FORK_EXEC_CMD, NULL},
     {NULL},
     {NULL}
    },
    {"ar7", "acroread", USE_SENDX,
     AR7_WIN_NAME, AR7_WIN_NAME_FMT, 
     {NULL},
     {NULL},
     {"Ctrl-W", "Alt-Left", NULL},
     {NULL},
     {NULL}
    },
    {"ar5", "acroread", USE_SENDX,
     AR5_WIN_NAME, AR5_WIN_NAME_FMT, 
     {NULL},
     {NULL},
     {"Ctrl-W", "Alt-Left", NULL},
     {NULL},
     {NULL}
    },
    /* 
     * xpdf also allows 'xpdf -remote tex -reload', but if that xpdf server
     * is not running, rather than a failure return, a blank window is
     * opened.  This is not much good to us.
     * Using "tex-server" as the server name is an arbitrary choice.
     */
    {"xpdf", "xpdf", USE_SENDX,
     XPDF_WIN_NAME, XPDF_WIN_NAME_FMT,
     {"-remote", "tex-server", NULL},
     {NULL},
     {"r", NULL},
     {NULL},
     {"%d", NULL}
    },
    {"evince", "evince", USE_SENDX,
     EVINCE_WIN_NAME, EVINCE_WIN_NAME_FMT,
     {NULL},
     {NULL},
     {"Ctrl-R", NULL},
     {"-i", "%d", NULL},
     {NULL}
    },
#ifdef TRY_OCULAR
    {"okular", "okular", USE_SENDX,
     OKULAR_WIN_NAME, OKULAR_WIN_NAME_FMT,
     {NULL},
     {NULL},
     {FOCUS_IN, "Fn-F5", FOCUS_OUT, NULL},
     {NULL},
     {NULL}
    },
#endif
};

#define     DEFAULT_VIEWER	(pdf_viewers[0].short_name)



static void
usage(void)
{
    int i;
    int viewers = sizeof(pdf_viewers) / sizeof(pdf_viewers[0]);

    fprintf(stderr, "This is version %s of %s.\n", VERSION, progname);
    fprintf(stderr, "Usage:\n  %s [-h|--help]\n", progname);
    fprintf(stderr, "    Show this help and exit.\n");

    fprintf(stderr, "  %s [-v|--version]\n", progname);
    fprintf(stderr, "    Show the version number and exit.\n");
    
    fprintf(stderr, "  %s [-r|--reset_focus] [-viewer <prog>] [-p <n>] "
	    "<file.pdf>\n", progname);
    fprintf(stderr, "    If the PDF viewer <prog> is displaying <file.pdf>, "
	    "reload that file.\n");
    fprintf(stderr, "    Otherwise call <prog> to display <file.pdf>.\n");
    fprintf(stderr, "    If '-r' or '--reset_focus' is used, attempt to");
    fprintf(stderr, " reset the input focus\n");
    fprintf(stderr, "    after calling the viewer program.\n");
    fprintf(stderr, "    If the viewer allows it and is not already\n");
    fprintf(stderr, "    running, the argument of -p specifies the\n");
    fprintf(stderr, "    first page to display.\n");
    fprintf(stderr, "    The default viewer program is %s.\n", DEFAULT_VIEWER);
    fprintf(stderr, "    Implemented viewers:");
    for (i = 0; i < viewers - 1; i++)
	fprintf(stderr, " '%s'", pdf_viewers[i].short_name);
    fprintf(stderr, " and '%s'.\n", pdf_viewers[viewers - 1].short_name);
    fprintf(stderr, "    Note: ar9 uses acroread's -openInNewInstance argument,"
	    	    " ar9-tab does not.\n");
}



/*
 * Send one "command" to the given window.
 * The command can be of these forms (for X an upper case letter or arrow
 * key, such as "Left", and D one or more digits; see sendx.c):
 * - Ctrl-X
 * - Alt-X
 * - Ctrl-Atl-X
 * - Fn-FD
 * - FOCUS-[IN|OUT]
 * - <string>	(not starting with any of the above prefixes)
 * Allow other variations on the cases of "Ctrl" and "Alt".
 * 
 * Return 0 on success, non-0 otherwise.
 */

static int
send_command(const char * window_name, const char * reload_cmd)
{
    int alt_len, ctrl_len, ctrl_alt_len, fn_len, focus_len;

    if (!strcmp(reload_cmd, FORK_EXEC_CMD))
	return 999;

    ctrl_alt_len = strlen(CTRL_ALT);
    if (!strncasecmp(CTRL_ALT, reload_cmd, ctrl_alt_len))
	return sendx_controlalt_token(reload_cmd + ctrl_alt_len, window_name);

    ctrl_len = strlen(CTRL);
    if (!strncasecmp(CTRL, reload_cmd, ctrl_len))
	return sendx_control_token(reload_cmd + ctrl_len, window_name);

    alt_len = strlen(ALT);
    if (!strncasecmp(ALT, reload_cmd, alt_len))
	return sendx_alt_token(reload_cmd + alt_len, window_name);

    fn_len = strlen(FN);
    if (!strncasecmp(FN, reload_cmd, fn_len))
	return sendx_token(reload_cmd + fn_len, window_name);

    focus_len = strlen(FOCUS_IN);
    if (!strncasecmp(FOCUS_IN, reload_cmd, focus_len))
	return set_focus(window_name);

    focus_len = strlen(FOCUS_OUT);
    if (!strncasecmp(FOCUS_OUT, reload_cmd, focus_len))
	return reset_focus();

    return sendx_string(reload_cmd, window_name);
}



/*
 * Try to reload the file.
 * If not successful, try to start up a new instance of the viewer program.
 * Return 0 on success, non-0 otherwise.
 * 
 * If reset_focus is true, (attempt to) reset the input focus to the
 * window which had focus when this program was called.  This is
 * currently only implemented for the USE_SENDX case.
 */

static int
view_file(const char * filename, int viewer_index, int focus_reset,
	  int initial_page)
{
    char * window_name;
    pid_t pid;
    int	success = 0;
    pdf_viewer_t pdf_viewer = pdf_viewers[viewer_index];
    const char ** reload_cmds = pdf_viewer.reload_cmds;

    /* First try to reload the file */
    switch (pdf_viewers[viewer_index].method)
    {
      case USE_SENDX:
        window_name = make_window_name(pdf_viewer.window_name_fmt, filename);
	if (window_name == NULL)
	    return 99;
	if (focus_reset)
	    (void)set_focus(window_name);
	while (*reload_cmds)
	{
	    /*
	     * Note: there is a race condition in the sendx.c code which
	     * can cause the program to crash and burn with a 'BadWindow'
	     * fault.  This has been observed for AR8 (but not AR7, oddly)
	     * after the ^W causes AR8 to change the window name.  It seems
	     * that Window_With_Name() tries to access a BadWindow within
	     * its recursive call.
	     * If this continues to be a problem, a usleep() may make the
	     * problem go away.
	     */
	    success = ! send_command(window_name, *reload_cmds);
	    if (! success && reload_cmds != pdf_viewer.reload_cmds)
	    {
		/*
		 * In some cases the window name might be changed because
		 * of the first command.
		 * For example, in AR7 the first command (^W) "closes"
		 * the document, and the window title changes to reflect
		 * the fact there is no document.  So try that command with
		 * the "plain" window name.
		 */
		success = ! send_command(pdf_viewer.window_name, *reload_cmds);
	    }
	    if (! success)
		break;
	    reload_cmds++;
	}
	if (focus_reset)
	    (void)reset_focus();
	break;

      case USE_FORK_EXEC:
	return 2;  /* TODO: someone who cares can put some code here */
	break;

      default:
	fprintf(stderr, "%s: internal error: invalid reload_method.  Sorry.\n",
		progname);
	return 99;
	break;
    }

    if (success)
	return 0;

    /* 
     * No luck re-loading the file.
     * (Try to) start a new instance of the viewer program.
     */
    if ((pid = fork()) == 0)
    {
	const char * argv_list[2 * MAX_OPTIONS + 3];
	const char ** p;
	int i = 0;

	argv_list[i++] = pdf_viewer.program_name;
	for (p = pdf_viewer.exec_pre_options; *p;)
	    argv_list[i++] = *p++;

	if (pdf_viewer.initial_page_pre_options[0] != NULL)
	{
	    int j = 0;
	    char * buf;

	    while (pdf_viewer.initial_page_pre_options[j + 1] != NULL)
		argv_list[i++] = pdf_viewer.initial_page_pre_options[j++];

	    /* 10 is more than enough for the page number */
	    buf = malloc(strlen(pdf_viewer.initial_page_pre_options[j]) + 10);
	    if (buf != NULL)
	    {
		/* 1e8 or more pages?  I doubt it. */
		sprintf(buf, pdf_viewer.initial_page_pre_options[j],
			initial_page > 99999999 ? 1 : initial_page);
		argv_list[i++] = buf;
	    }
	    else
		fprintf(stderr, "%s: unable to allocate memory for page "
			"specification!!\n"
			"Crossing fingers and continuing.", progname);
	}

	argv_list[i++] = filename;

	if (pdf_viewer.initial_page_post_options[0] != NULL)
	{
	    int j = 0;
	    char * buf;

	    while (pdf_viewer.initial_page_post_options[j + 1] != NULL)
		argv_list[i++] = pdf_viewer.initial_page_post_options[j++];
	    
	    buf = malloc(strlen(pdf_viewer.initial_page_post_options[j]) + 10);
	    if (buf != NULL)
	    {
		sprintf(buf, pdf_viewer.initial_page_post_options[j],
			initial_page > 99999999 ? 1 : initial_page);
		argv_list[i++] = buf;
	    }
	    else
		fprintf(stderr, "%s: unable to allocate memory for page "
			"specification!!\n"
			"Crossing fingers and continuing.", progname);
	}

	argv_list[i++] = NULL;
#ifdef DEBUG
	{
	    int k = 0;
	    for (; k < i-1; k++)
		fprintf(stderr, "%s\n", argv_list[k]);
	}
#endif
	if (execvp(pdf_viewer.program_name, (char **)argv_list))
	{
	    fprintf(stderr, "%s: %s startup failed\n", progname,
		    pdf_viewer.program_name);
	    return 5;
	}
    }
    else if (pid < 0)
    {
	fprintf(stderr, "%s: fork() failed!?\n", progname);
	return 10;
    }
    
    return 0;
}



int
main(int argc, char * argv[])
{
    const char * viewer = pdf_viewers[0].short_name;
    char * env_viewer;
    int i;
    int initial_page = 1;
    int viewer_index = -1;
    int focus_reset = 0;

    progname = argv[0];
    argc--;
    argv++;

    env_viewer = getenv("PDF_VIEWER");
    if (env_viewer != NULL)
	viewer = env_viewer;

    if (argc < 1 || !strcmp(argv[0], "-h") || !strcmp(argv[0], "--help"))
    {
	usage();
	return argc < 2 ? EXIT_FAILURE : EXIT_SUCCESS;
    }

    if (!strcmp(argv[0], "-v") || !strcmp(argv[0], "--version"))
    {
	printf("This is pdfopen version %s\n", VERSION);
	return EXIT_SUCCESS;
    }

    while (argc > 0)
    {
	if (!strcmp(argv[0], "--reset_focus") || !strcmp(argv[0], "-r"))
	{
	    argc--;
	    argv++;
	    focus_reset = 1;
	    continue;
	}

	if (argc >= 2 && !strcmp(argv[0], "-viewer"))
	{
	    viewer = argv[1];
	    argc -= 2;
	    argv += 2;
	    continue;
	}

	if (argc >= 2 && !strcmp(argv[0], "-p"))
	{
	    initial_page = atoi(argv[1]);
	    argc -= 2;
	    argv += 2;
	    continue;
	}

	break;
    }

    /* Should only have 1 filename left now */
    if (argc != 1)
    {
	fprintf(stderr, "%s: invalid argument(s)\n", progname);
	usage();
	return EXIT_FAILURE;
    }

    for (i = 0; i < sizeof(pdf_viewers) / sizeof(pdf_viewers[0]); i++)
    {
	if (!strcmp(pdf_viewers[i].short_name, viewer))
	{
	    viewer_index = i;
	    break;
	}
    }

    if (viewer_index < 0)
    {
	fprintf(stderr, "%s: Unknown PDF viewer '%s'\n", progname, viewer);
	usage();

	return EXIT_FAILURE;
    }

    if (view_file(argv[0], viewer_index, focus_reset, initial_page))
	return EXIT_FAILURE;

    return EXIT_SUCCESS;
}
